File: Java/src/com/codewith/mosh/LinkedList.java
package com.codewith.mosh;

import java.util.NoSuchElementException;

public class LinkedList {
    private class Node {
        private int value;
        private Node next;
        Node(int value) {
            this.value = value;
        }
    }
    private Node first;
    private Node last;
    private int size;
    // private int[] list;
    private boolean isEmpty(){
        return first == null;
    }
    private Node getPrevious(Node item) {
        var currentNode = first;
        while(currentNode != null) {
            if (currentNode.next == item) return currentNode;
            currentNode = currentNode.next;   
        } 
        return null;
    }
    public void addLast(int value) {
        var node = new Node(value);
        if (isEmpty()) {
            first = last = node;
        } else {
            last.next = node;
            last = node;
        }
        size++;
    }
    public void addFirst(int value) {
        var node = new Node(value);
        if (isEmpty()){
            first = last = node;
        }
        else {
            node.next = first;
            first = node;
        }
        size++;
    }
    public int indexOf(int item) {
        int index = 0;
        // boolean found = false;
        var currentNode = first;
        while (currentNode != null) {
            if (currentNode.value == item) return index;
            currentNode = currentNode.next;
            index++;
        }
        return -1;
    }
    public boolean contains(int item) {
        return indexOf(item) != -1;
    }
    public void removeFirst() {
        if (isEmpty()) {
            throw  new NoSuchElementException();
        }
        if (first == last) {
            first = last = null;
        } else {
        var second = first.next;
        first.next = null;
        first = second;
        }

        size--;
    }
    public void removeLast() {
        if (isEmpty()) {
            throw  new NoSuchElementException();
        }
        if (first == last) {
            first = last = null;
        } else {
            var previousNode = getPrevious(last);
            last = previousNode;
            last.next = null;
        }
        size--;
    }
    public int size() {
        return size;
    }
    public int[] toArray() {
        int[] array = new  int[size];
        var current = first;
        var index = 0;
        while (current != null ) {
            array[index++] = current.value;
            current = current.next;
        }
        return array;
    }
    public void reverse(){
        if (isEmpty()) {
            return;
        }
        var previous = first;
        var current = first.next;
        while (current != null){
            var next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        last = first;
        last.next = null;
        first = previous;
    }
    public int findKthFromEnd(int k) {
        if (isEmpty()) {
            throw  new  IllegalStateException();
        }
        var p1 = first;
        var p2 = first;
        if (k == 0) {
            throw  new  IllegalArgumentException();
        } else if (k > size) {
            return -1;
        } else if (k == size) {
            return first.value;
        }
        for (var i=0; i< k-1; i++) {
            p2 = p2.next;
            if (p2 == null) {
               throw  new  IllegalArgumentException(); 
            }
        }
        while(p2 != last) {
            p1 = p1.next;
            p2 = p2.next;
        }
        return p1.value;
    }
}

File: Python/Exercises/linkedlist.py
from ds import ListNode
class Node:
    def __init__(self, value):
        self.value =value
        self.next = None
    
    def __eq__(self, node):
        if isinstance(node, Node):
            return False
        return self.value == node.value

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    
    def _isEmpty(self):
        return self.head == self.tail == None
    
    def _getPreviousNode(self, node: Node):
        current = self.head
        while current.next:
            if current.next == node:
                return current.value
            current = current.next
        return None
    def addFirst(self, value: int) -> None:
        node = Node(value=value)
        if self._isEmpty():
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node
            self.size += 1
    
    def addLast(self, value: int) -> None:
        node = Node(value=value)
        if self._isEmpty():
            self.head = self.tail = node
        else:
            prev = self.tail
            prev.next = node
            self.tail = node
            self.size += 1
    
    def removeFirst(self) -> None:
        if self._isEmpty():
            raise Exception('Linked List is empty')
        else:
            new_head = self.head.next
            self.head = new_head 
            self -= 1
    

    def removeLast(self):
        if self._isEmpty():
            raise Exception('Linked List is empty')
        else:
            previous = self._getPreviousNode(self.tail)
            previous.next = None
            self.head = previous
    
    def reverse(self):
        reversedll = LinkedList()
        current = self.head
        while current:
            reversedll.addFirst(current.value)
            current = current.next
        return reversedll

    def toArray(self) -> list:
        current = self.head
        list = []
        while current:
            print(current.value)
            list.append(current.value)
            current = current.next
        return list


ll = LinkedList()
ll.addFirst(10)
ll.addFirst(20)
ll.addFirst(30)
ll.addFirst(40)
ll.addLast(50)
ll.addLast(60)
ll.addLast(70)

k = 2 
if ll.head 
# print("reversed")
# reversed.toArray()
# ll.removeLast()
# ll.removeFirst()
# ll.removeFirst()
# ll.removeFirst()
# print("normal")
# ll.toArray()
